### PR TITLE
Add a strain abbreviation column and synonyms

### PR DESCRIPTION
**Pull request by kimrutherford**
Hi James.  I've done the schema change but I haven't written the code to update an existing database.  I'll do that tomorrow.  I'll let you know when there's something to test.

Refs #2231
Refs #2232

### CODE REVIEW

**Comment by kimrutherford** [comment]
> I've done the schema change but I haven't written the code to update an existing database. 

I've updated the upgrade script to handle this schema change.  There's not much to test yet because nothing makes use of the new column and table.

**Comment by kimrutherford** [comment]
Hi James.

The loading code will now populate the strainsynonym table from the input file.  eg.

```csv
 NcbiTaxSpeciesId,CommonName,StrainCultivar
 3702,thale cress,Columbia Col-0,COL0,"col-0,columbia-0,colcol-0"
 9606,human,CFBE41o cells
```

If there are multiple strains that field will need to be quoted.

As we discussed, the strain details and its synonyms are completely reloaded each time.  So any existing abbreviation and synonyms will be removed and replaced with what is in the file.  If the abbreviation column is empty, the abbreviation will be removed from the strain.  If the synonyms column is empty all existing synonyms will be removed.

If a strain is completely absent from the input file the strain will be left untouched in the database.  So cuurently there is no way to delete a strain.  We can talk about that.

I've updated the wiki: https://github.com/pombase/canto/wiki/StrainDataFile

I think this code is now OK enough for you to test when you get a chance.  You'll need to run upgrade_db.pl

**Comment by kimrutherford** [comment]
After discussion with James, I've removed the "strain_abbreviation" column.  We'll add it back in future if it's needed.

**Comment by kimrutherford** [comment]
Hi James.  I think this is ready to merge.  Would you be able to test upgrading the schema?

**Comment by jseager7** [comment]
Sorry about the delay on this. I've tested the schema upgrade itself and that worked with no problems on my local copy. I haven't tested it on the production database yet, though.

**Comment by jseager7** [comment]
Spoke too soon, I get an error when navigating to the edit_genes page:

`
DBIx::Class::Storage::DBI::_prepare_sth(): DBI Exception: DBD::SQLite::db prepare_cached failed: no such column: strains.sraing [for Statement "SELECT strains.strain_id, strains.organism_id, strains.strain_name, strains.sraing FROM organismprop me JOIN cvterm type ON type.cvterm_id = me.type_id JOIN organism organism ON organism.organism_id = me.organism_id JOIN strain strains ON strains.organism_id = organism.organism_id WHERE ( ( type.name = ? AND value = ? ) )"] at lib/Canto/Track/StrainLookup.pm line 93
`

Surely `strains.sraing` is a typo?

File: lib/Canto/TrackDB/Strain.pm [hunk 1]
```diff
@@ -58,8 +57,8 @@ __PACKAGE__->add_columns(
   { data_type => "integer", is_foreign_key => 1, is_nullable => 0 },
   "strain_name",
   { data_type => "text", is_nullable => 0 },
-  "strain_abbreviation",
-  { data_type => "text\n       sraing", is_nullable => 1 },
+  "sraing",
```
Comment (jseager7) [inline lib/Canto/TrackDB/Strain.pm:60]: I think this is where the "sraing" value first appears. I'm guessing this isn't intentional?

**Comment by kimrutherford** [comment]
> Surely strains.sraing is a typo?

Whoops!  Thanks for spotting that.  I've pushed a fix.

**Comment by kimrutherford** [comment]
I've just pushed a fix to allow non-ASCII characters (like delta symbols) in strain names.

**Comment by jseager7** [comment]
Thanks. I'll test the loading again tomorrow.

**Comment by jseager7** [comment]
I've tested the loading and it seems all the strains loaded successfully this time. Thanks!

I've also checked them in Canto, and while we don't have any way to display the strain synonyms yet, it seems everything else still works fine. Seems the encoding issues are fixed as well, based on the lowercase Chi (χ) in this example:

image

I had one problem related to an obsolete taxon ID while doing this, which I've opened a new issue about at #2264.

I don't think there's anything left to stop this from merging.

**Comment by jseager7** [comment]
Fixes #2233

**Comment by jseager7** [comment]
Oh, I didn't realise you'd already done the required service changes for strain synonyms in https://github.com/pombase/canto/issues/2236. Let me know if you want me to test those changes too before merging this.

**Comment by kimrutherford** [comment]
> Let me know if you want me to test those changes too before merging this.

I think we can go ahead and merge then fix problems later.   Thanks for all the testing.

**Pull request merged**